### PR TITLE
Feature lwdev 5904 overnight swaps clear

### PR DESCRIPTION
**Pull request by GSailen**

### CODE REVIEW

File: src/MarginTrading.AzureRepositories/Entities/OvernightSwapHistoryEntity.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		
+		public static string GetKey(string accountId, string instrument, OrderDirection? direction) =>
+			$"{accountId}_{instrument ?? ""}_{direction?.ToString() ?? ""}";
+
+		public static OvernightSwapHistoryEntity Create(IOvernightSwapHistory obj)
+		{
+			return new OvernightSwapHistoryEntity
+			{
+				PartitionKey = obj.AccountId,
+				RowKey = GetKey(obj.AccountId, obj.Instrument, obj.Direction),
```
Comment (andrey-anokhin) [inline src/MarginTrading.AzureRepositories/Entities/OvernightSwapHistoryEntity.cs]: PartitionKey + RowKey does not depend on date and not good idea for history. I know that in repo you generate PK, but it should be removed from here.
Copy-paste? )

File: src/MarginTrading.AzureRepositories/OvernightSwapStateRepository.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		{
+			await _tableStorage.InsertOrReplaceAsync(OvernightSwapStateEntity.Create(obj));
+		}
+
+		public async Task<IEnumerable<IOvernightSwapState>> GetAsync()
+		{
+			return await _tableStorage.GetDataAsync();
+		}
+
+		public async Task<IReadOnlyList<IOvernightSwapState>> GetAsync(string accountId, DateTime? @from, DateTime? to)
```
Comment (andrey-anokhin) [inline src/MarginTrading.AzureRepositories/OvernightSwapStateRepository.cs]: Not used and no idea how can be used.

File: src/MarginTrading.Backend.Contracts/DataReaderClient/MtDatareaderClient.cs [hunk 3]
```diff
@@ -23,6 +25,7 @@ public MtDataReaderClient(string url, string apiKey, string userAgent)
             AccountHistory = RestService.For<IAccountHistoryApi>(url, settings);
             AccountsApi = RestService.For<IAccountsApi>(url, settings);
             TradeMonitoringRead = RestService.For<ITradeMonitoringReadingApi>(url, settings);
+            TradingConditionsRead = RestService.For<ITradingConditionsReadingApi>(url, settings);
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Contracts/DataReaderClient/MtDatareaderClient.cs]: AccountAssetPairsRead init is missed

File: src/MarginTrading.Backend.Core/Caches/IOvernightSwapCache.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+	{
+		OvernightSwapCalculation Get(string key);
+		bool TryGet(string key, out OvernightSwapCalculation item);
+		IReadOnlyList<OvernightSwapCalculation> GetAll();
+		bool TryAdd(OvernightSwapCalculation item);
+		bool AddOrReplace(OvernightSwapCalculation item);
+		void SetAll(IEnumerable<OvernightSwapCalculation> items);
+		void Clear(string key);
+		void ClearAll();
+		void Initialize(IEnumerable<OvernightSwapCalculation> items);
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Core/Caches/IOvernightSwapCache.cs]: Difference between Initialize and SetAll?

File: src/MarginTrading.Backend.Core/OvernightSwapCalculation.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace MarginTrading.Backend.Core
+{
+	public class OvernightSwapCalculation : IOvernightSwapHistory, IOvernightSwapState
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Core/OvernightSwapCalculation.cs]: Why need both if IOvernightSwapHistory : IOvernightSwapState ?

File: src/MarginTrading.Backend.Core/Services/IQuoteCacheService.cs [hunk 6]
```diff
@@ -6,6 +6,7 @@ public interface IQuoteCacheService
     {
         InstrumentBidAskPair GetQuote(string instrument);
         Dictionary<string, InstrumentBidAskPair> GetAllQuotes();
-        bool TryGetQuoteById(string instrument, out InstrumentBidAskPair result);
+        bool TryGetQuoteById(string instrument, out InstrumentBidAskPair result); 
+        void RemoveQuote(string assetPair);
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Core/Services/IQuoteCacheService.cs]: This is from other task I think, but ok )

File: src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+		public OvernightSwapCache()
+		{
+			ClearAll();
+		}
+        
+		public OvernightSwapCalculation Get(string key)
+		{
+			lock (LockObj)
+			{
+				return _cache.TryGetValue(key, out var value)
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs]: Double lock?

File: src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		}
+
+		public bool TryAdd(OvernightSwapCalculation item)
+		{
+			if (item == null || _cache.ContainsKey(item.Key))
+				return false;
+            
+			lock (LockObj)
+			{
+				_cache[item.Key] = item;
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs]: _cache.ContainsKey(item.Key) should be checked 2nd time inside lock

File: src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+			lock (LockObj)
+			{
+				_cache.Remove(item.Key);
+				_cache[item.Key] = item;
+			}
+
+			return true;
+		}
+
+		public void SetAll(IEnumerable<OvernightSwapCalculation> items)
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs]: May be it is better to make it private? It is used only in Initialize

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+using Lykke.Common;
+using MarginTrading.Backend.Core;
+using MarginTrading.Backend.Core.Repositories;
+using MarginTrading.Backend.Core.Settings;
+using MarginTrading.Backend.Core.TradingConditions;
+using MarginTrading.Backend.Services.Assets;
+using MarginTrading.Backend.Services.TradingConditions;
+using Newtonsoft.Json;
+
+namespace MarginTrading.Backend.Services
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: Rider highlights namespace )

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+			CalculateAndChargeSwaps();
+			
+			//TODO if server was down more that a day.. calc N days
+		}
+
+		public void CalculateAndChargeSwaps()
+		{
+			lock (LockObj)
+			{
+				_currentStartTimestamp = DateTime.UtcNow;
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: It is better to use IDateService for getting current time

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+				//read orders syncronously
+				var openOrders = _orderReader.GetActive();
+
+				//filter orders that are already calculated
+				var calculatedIds = _overnightSwapCache.GetAll()
+					.Where(x => x.Time > LastInvocationTime)
+					.SelectMany(x => x.OpenOrderIds);
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: It is better to calculate LastInvocationTime once and transform result to HashSet, as it is used only for Contains (IEnumerable is the worst option in that case).

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+			
+			//start calculation
+			CalculateAndChargeSwaps();
+			
+			//TODO if server was down more that a day.. calc N days
+		}
+
+		public void CalculateAndChargeSwaps()
+		{
+			lock (LockObj)
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: Lock will not work here, as calculation is processed in new thread. It is better to use flag.. or may be lock inside calc function (for case when server is started just before new calculation should be processed).
And.. lock before getting orders is bad idea anyway.

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+		/// Calculate overnight swaps for account/instrument/direction order package.
+		/// </summary>
+		/// <param name="instrument"></param>
+		/// <param name="account"></param>
+		/// <param name="accountAssetPair"></param>
+		/// <param name="direction"></param>
+		/// <param name="orders"></param>
+		/// <returns></returns>
+		private async Task ProcessOrders(IReadOnlyList<Order> orders, string instrument, MarginTradingAccount account,
+				IAccountAssetPair accountAssetPair, OrderDirection direction)
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: It is better to use domain interfaces, not DTO. E.g. MarginTradingAccount -> IMarginTradingAccount

File: src/MarginTrading.Backend/Application.cs [hunk 15]
```diff
@@ -93,11 +96,16 @@ public sealed class Application
...
+                var registry = new Registry();
+                registry.Schedule<OvernightSwapJob>().ToRunEvery(0).Days().At(_marginSettings.OvernightSwapCalculationHour, 0);
+                JobManager.Initialize(registry);
+                JobManager.JobException += info => _logger.WriteError(ServiceName, nameof(JobManager), info.Exception);
             }
             catch (Exception ex)
             {
                 _consoleWriter.WriteLine($"{ServiceName} error: {ex.Message}");
-                await _logger.WriteErrorAsync(ServiceName, nameof(StartApplicationAsync), null, ex);
+                await _logger.WriteErrorAsync(ServiceName, "Application.RunAsync", null, ex);
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend/Application.cs]: Why not nameof? )

File: src/MarginTrading.Backend/Controllers/OvernightSwapController.cs [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+	public class OvernightSwapController : Controller
+	{
+		private readonly IOvernightSwapHistoryRepository _overnightSwapHistoryRepository;
+		
+		public OvernightSwapController(IOvernightSwapHistoryRepository overnightSwapHistoryRepository)
+		{
+			_overnightSwapHistoryRepository = overnightSwapHistoryRepository;
+		}
+		
+		[Route("overnight.swaps.state")]
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend/Controllers/OvernightSwapController.cs]: Strange route api/overnightswap/overnight.swaps.state. 
And it is not state, but history.. 
May be change to [Route("history")] ?

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+				_currentStartTimestamp = DateTime.UtcNow;
+
+				//read orders syncronously
+				var openOrders = _orderReader.GetActive();
+
+				//filter orders that are already calculated
+				var calculatedIds = _overnightSwapCache.GetAll()
+					.Where(x => x.Time > LastInvocationTime)
+					.SelectMany(x => x.OpenOrderIds);
+				var filteredOrders = openOrders.Where(x => !calculatedIds.Contains(x.Id));
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: And what about already closed orders?
If last calculation failed and user closed order? We need to find them in history or, at least send critical error to slack.

File: src/MarginTrading.AzureRepositories/OvernightSwapHistoryRepository.cs [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+		public async Task<IEnumerable<IOvernightSwapHistory>> GetAsync()
+		{
+			return await _tableStorage.GetDataAsync();
+		}
+
+		public async Task<IReadOnlyList<IOvernightSwapHistory>> GetAsync(DateTime? @from, DateTime? to)
+		{
+			var partitionKeys = await GetPartitionKeys();
+			return (await _tableStorage.WhereAsync(partitionKeys, from ?? DateTime.MinValue, to ?? DateTime.MaxValue, ToIntervalOption.IncludeTo))
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AzureRepositories/OvernightSwapHistoryRepository.cs]: Cool, used that solution.

File: src/MarginTrading.AzureRepositories/Entities/OvernightSwapHistoryEntity.cs (hunk 1 above)
Comment (GSailen) [reply src/MarginTrading.AzureRepositories/Entities/OvernightSwapHistoryEntity.cs]: That was copy-paste, yep. Fixed.

File: src/MarginTrading.AzureRepositories/OvernightSwapStateRepository.cs (hunk 2 above)
Comment (GSailen) [reply src/MarginTrading.AzureRepositories/OvernightSwapStateRepository.cs]: Removed.

File: src/MarginTrading.Backend.Core/Caches/IOvernightSwapCache.cs (hunk 4 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Core/Caches/IOvernightSwapCache.cs]: No difference in this specific case. Removed unused methods from interface.

File: src/MarginTrading.Backend.Core/OvernightSwapCalculation.cs (hunk 5 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Core/OvernightSwapCalculation.cs]: To make code more readable.

File: src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs (hunk 8 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs]: Good idea, fixed.

File: src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs (hunk 9 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Services/Caches/OvernightSwapCache.cs]: +

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs (hunk 10 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: +

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs (hunk 11 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: Done

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs (hunk 12 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: Done

File: src/MarginTrading.Backend/Application.cs (hunk 15 above)
Comment (GSailen) [reply src/MarginTrading.Backend/Application.cs]: Must be cherry-picked from Nuno's commit, so can't tell why.

File: src/MarginTrading.Backend/Controllers/OvernightSwapController.cs (hunk 16 above)
Comment (GSailen) [reply src/MarginTrading.Backend/Controllers/OvernightSwapController.cs]: Fixed

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs (hunk 13 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: Fixed. Used async lock inside calc method.

File: src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs (hunk 17 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs]: Detection of such cases added.

**Pull request merged**